src/plugin-traverser.ts

```diff
@@ -0,0 +1,115 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+import { BasePlugin } from './abstract/base-plugin.plugin';
+import { PluginModuleOptions } from './interfaces/plugin-module-options.interface';
+import { PluginOptions } from './interfaces/plugin-options.interface';
+import { Provider } from '@nestjs/common';
+
+const node_modules = require('node_modules-path');
+
+export class PluginTraverser {
+  private _pluginType: string;
+  private _directories: Array<string> = [path.resolve(node_modules())];
+
+  constructor(pluginModuleOptions: PluginModuleOptions) {
+    this._pluginType = pluginModuleOptions.pluginType;
+    if (
+      pluginModuleOptions.directories &&
+      pluginModuleOptions.directories.length > 0
+    )
+      this._directories = pluginModuleOptions.directories;
+  }
+
+  public traverseDirectories(): Array<Provider<BasePlugin>> {
+    let modules = [];
+    this._directories.forEach((parentDirectory) => {
+      if (fs.existsSync(parentDirectory)) {
```

Comment:
We can use [Promise.each](http://bluebirdjs.com/docs/api/promise.each.html) to read folders asyncronous.